New API methods for export files

AgustinSilverfin opened the pull request:
## Description

Include a summary of the changes made

Fixes # (link to the corresponding issue if applicable)

## Type of change

- [ ] Bug fix
- [X] New feature
- [ ] Breaking change

## Checklist

- [ ] README updated (if needed)
- [X] Version updated (if needed)
- [ ] Documentation updated (if needed)

Woetfin commented:
Shouldn't we split this up in separate PRs?

Maybe we can then release each part in this order and already confirm that those parts are running well: 
1. Refactor templates into Classes. Align method names.
2. Add support for account details.
3. Add support for export files.
4. Add --message flag for update and create commands

AgustinSilverfin commented:
I can split it like that as long as the refactoring is the first one :slightly_smiling_face: 
So, let's focus on getting this done first #55 
And then I'll split this one into separate PRs

Woetfin commented:
@AgustinSilverfin Sounds good! 

I think this request might also be a good one to pick up during the refactoring? 
https://github.com/silverfin/silverfin-cli/issues/37

Woetfin commented on index.js:
```diff
@@ -111,6 +113,111 @@ async function newAllReconciliations(firmId) {
+  }
+}
+
+async function fetchAccountDetailByHandle(firmId, name) {}
+async function fetchAccountDetailById(firmId, id) {}
+async function fetchAllAccountDetails(firmId) {}
+async function publishAccountDetailByName(firmId, name) {}
+async function publishAllAccountDetails(firmId) {}
+async function newAccountDetail(firmId, name) {}
+async function newAllAccountDetails(firmId) {}
```
Should we leave these as part of this PR?

AgustinSilverfin replied:
Removed

Woetfin commented on index.js:
```diff
@@ -211,114 +318,99 @@ async function newAllSharedParts(firmId) {
 }
 
-/** This function adds a shared part to a reconciliation text. It will make a POST request to the API. If the ID of one of the templates it missing, it will try to fetch it first by making a GET request. In case of success, it will store the details in the corresponding config files.
+/** This function adds a shared part to a template. It will make a POST request to the API. If the ID of one of the templates it missing, it will try to fetch it first by making a GET request. In case of success, it will store the details in the corresponding config files.
```
```suggestion
/** This function adds a shared part to a template. It will make a POST request to the API. If the ID of one of the templates is missing, it will try to fetch it first by making a GET request. In case of success, it will store the details in the corresponding config files.
```

small typo

Woetfin commented on index.js:
```diff
@@ -327,65 +419,86 @@ async function addSharedPartToReconciliation(
+async function addAllSharedParts(firmId) {
   const sharedPartsArray = fsUtils.getAllTemplatesOfAType("sharedPart");
-  for (let sharedPartName of sharedPartsArray) {
+  for await (let sharedPartName of sharedPartsArray) {
```
I think the for .. of loop is already synchronous, so the await keyword isn't necessary

AgustinSilverfin replied:
I think we actually needed. Not sure if both awaits :thinking:

Woetfin replied:
I did a test and I believe it's finishing the first loop and doing it in sequential order, so I don't think the await is necessary here; checking this documentation https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of ; I believe it still works cause we're using it to loop over 'sync iterables', but we can do that as well with the normal one: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of

AgustinSilverfin replied:
Removed

Woetfin reviewed (changes requested):
Liking the organisation of the template types in separate classes 👍 making we can add those to a separate "templateTypes" directory as well inside the "lib" directory? 

Tested all commands:
1. import-export-file
	1. Change to only support import with id so it's aligned with the account templates? 
	2. function checkValidName ==> for export files / account templates, spaces and other characters should be allowed since the name of the folder will be a name that could contain this and be valid. 
	3. the import also creates a "tests" folder, but since there are no liquid tests for export files, it can be skipped. 
2. update-export-file
	1. Looks good
3. development-mode
	1. Update from export file doesn't seem to trigger update from dev mode 
4. create-export-file
	1. the create also creates a "tests" folder, but since there are no liquid tests for export files, it can be skipped. 
5. add-shared-part
	1. Looks good
6. remove-shared-part
        1. Looks good
7. get-export-file-id
	1. Looks good

AgustinSilverfin commented:
@Woetfin I think I have addressed all your comments. Could you test again?
As we discussed, I kept the import with name as consistency (since you can update and create with name)

Woetfin reviewed (approved):
Tested commands for the export files import / update / add shared part / remove shared part / development mode and all working well, so I think it's good to go 🔥
